### PR TITLE
unit tests for the annotation related functions in pkg/util

### PR DESCRIPTION
**Pull request by vjayaramrh**
Signed-off-by: Vishwanath Jayaraman <vjayaram@redhat.com>

**- What this PR does and why is it needed**

The PR adds unit tests (code refactor is not the scope for this PR) for most of the functions defined in the following files:

    - pkg/util/node_annotations.go 
    - pkg/util/pod_annotation.go
    -pkg/util/subnet_annotations.go

**- Special notes for reviewers**

**- How to verify it**

Pull this PR into your workspace
Navigate to go-controller/pkg/util directory
execute below commands
go test -coverprofile cp.out
go tool cover -html=cp.out
select files node_annotations.go, pod_annotation.go, subnet_annotations.go from drop down to verify coverage

**- Description for the changelog**

### CODE REVIEW

**Comment by vjayaramrh** [comment]
@trozet @danwinship would appreciate your review

File: go-controller/pkg/util/node_annotations_unit_test.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+	ovntest "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/testing"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	"net"
+	"reflect"
+	"testing"
+)
+import "fmt"
```
Comment (danwinship) [inline go-controller/pkg/util/node_annotations_unit_test.go:15]: the imports here are all messed up

File: go-controller/pkg/util/node_annotations_unit_test.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+			},
+			expOutput: []byte(`{"mode":"local","node-port-enable":"false"}`),
+		},
+		{
+			desc: "test VLANID not nil",
+			inpL3GwCfg: &L3GatewayConfig{
+				Mode:   config.GatewayModeLocal,
+				VLANID: &vlanid,
+			},
+			expOutput: []byte(`{"mode":"local","node-port-enable":"false","vlan-id":"1024"}`),
```
Comment (danwinship) [inline go-controller/pkg/util/node_annotations_unit_test.go:42]: This should actually be an error: `VLANID` is only allowed with shared gateway mode

File: go-controller/pkg/util/node_annotations_unit_test.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+			errExpected: true,
+			inputParam:  []byte(`{"mode":"local","mac-address":"11:22:33:44:55:66","ip-address":"192.168.1/24"}`),
+		},
+		{
+			desc:        "test bad 'IP addresses' value",
+			errExpected: true,
+			inputParam:  []byte(`{"mode":"local","mac-address":"11:22:33:44:55:66","ip-addresses":["192.168.1/24","fd01::1234/64"]}`),
+		},
+		{
+			desc:        "test bad 'next-hop value",
```
Comment (danwinship) [inline go-controller/pkg/util/node_annotations_unit_test.go:129]: (missing close single-quote)

File: go-controller/pkg/util/node_annotations_unit_test.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+			inputL3GwCfg:     L3GatewayConfig{},
+		},
+		{
+			desc:             "success: apply empty Chassis id ",
+			inpNodeAnnotator: testAnnotator,
+			inputL3GwCfg: L3GatewayConfig{
+				ChassisID: " ",
+			},
+		},
+	}
```
Comment (danwinship) [inline go-controller/pkg/util/node_annotations_unit_test.go:177]: This is clearly not a complete set of tests for `SetL3GatewayConfig`...

File: go-controller/pkg/util/node_annotations_unit_test.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+			inputParam: []byte(`{"mode":""}`),
+			expOut: L3GatewayConfig{
+				Mode:           "",
+				NodePortEnable: false,
+			},
+		},
+		{
+			desc:        "error: test unsupported gateway mode",
+			inputParam:  []byte(`{"mode":"blah"}`),
+			errExpected: true,
```
Comment (danwinship) [inline go-controller/pkg/util/node_annotations_unit_test.go:105]: It's generally better to assert more specific things about the error.

eg, if someone added a check that "mac-address" always has to be set, and then later broke the check for unsupported gateway mode, this test would still pass, because it would be returning a "mac address is unset" error, which is the wrong error, but would fool the test case.

in other places it's common to check for a substring match on the returned error

File: go-controller/pkg/util/subnet_annotations_unit_test.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+			res, err := CreateNodeLocalNatAnnotation(tc.inpIPList)
+			t.Log(res, err)
+			assert.Equal(t, tc.expOutput, res)
+		})
+	}
+}
+
+func TestSetNodeLocalNatAnnotation(t *testing.T) {
+
+}
```
Comment (danwinship) [inline go-controller/pkg/util/subnet_annotations_unit_test.go:171]: ?

**Comment by vjayaramrh** [comment]
> It still does not make sense to have two separate sets of unit tests. Either these tests should be merged into the existing ones, or the existing ones should be deleted and these ones extended to cover all of the things that the old ones tested as well

@trozet had made a comment regarding the approach on an earlier PR at https://github.com/ovn-org/ovn-kubernetes/pull/1382#issuecomment-648384112

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 1 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:15]: shall fix the imports

**Comment by danwinship** [comment]
> @trozet had made a comment regarding the approach on an earlier PR

> current approach by @vjayaramrh is to add unit tests via testify, and remove the leftover duplicates in ginkgo later. That way it will be more obvious if he removes duplicate tests in ginkgo in a separate PR, we can see coveralls change 0% and its easy merge.

So the new tests have 100% coverage on their own? They seem to be missing things that the old tests have. eg, the "`Nil entry in GW`" test in `pod_annotation_test.go`.

Even if the new tests have 100% line coverage, they may be missing specific code flows that were tested by the old tests...

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 2 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:42]: Thanks for catching this, however, I should be able to modify line 39 to be `Mode: config.GatewayModeLocal` and L42 to be `expOutput: []byte(`{"mode":"local","node-port-enable":"false","vlan-id":"1024"}`)` , right?

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 3 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:129]: Shall fix this

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 4 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:177]: Let me relook into the function code and see what I may have missed. Seems like I only have success cases currently.

File: go-controller/pkg/util/subnet_annotations_unit_test.go (hunk 6 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/subnet_annotations_unit_test.go:171]: Thanks for the catch, shall fix it.

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 2 above)
Comment (danwinship) [reply go-controller/pkg/util/node_annotations_unit_test.go:42]: Assuming that when you say `GatewayModeLocal` you mean `GatewayModeShared` and that when you say `"mode":"local"` you mean `"mode":"shared'`, then yes.

But also, I meant that this is pointing out a bug in the (non-test) code; it ought to be throwing an error in the case you have now.

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 5 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:105]: Valid point. If an error message is originating inside of the function being unit tested, checking for substring match would be effective.

**Comment by vjayaramrh** [comment]
> > @trozet had made a comment regarding the approach on an earlier PR
> 
> > current approach by @vjayaramrh is to add unit tests via testify, and remove the leftover duplicates in ginkgo later. That way it will be more obvious if he removes duplicate tests in ginkgo in a separate PR, we can see coveralls change 0% and its easy merge.
> 
> So the new tests have 100% coverage on their own? They seem to be missing things that the old tests have. eg, the "`Nil entry in GW`" test in `pod_annotation_test.go`.
> 
> Even if the new tests have 100% line coverage, they may be missing specific code flows that were tested by the old tests...

Below is the coverage from the newly written unit tests for node_annotations.go, pod_annotations.go and subnet_annotations.go
Screen Shot 2020-07-02 at 2 02 23 PM
I have tried to cover as much code as possible for the functions in those files. The coverage can easily be seen as HTML output by executing the below commands (rename the pod_annotation_test.go, node_annotations_test.go, subnet_annotations_test.go as pod_annotation_test.go.BAK, node_annotations_test.go.BAK, subnet_annotations_test.go.BAK  ) in the go-controller/pkg/util folder
```
go test -coverprofile cp.out
go tool cover -html=cp.out
```
Where feasible and with no code refactor, the unit tests try to cover as much code path as possible for the function being unit tested.

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 2 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:42]: Thanks for the correction again, Below should be the corrected test case
```
desc: "test VLANID not nil",
			inpL3GwCfg: &L3GatewayConfig{
				Mode:   config.GatewayModeShared,
				VLANID: &vlanid,
			},
			expOutput: []byte(`{"mode":"shared","node-port-enable":"false","vlan-id":"1024"}`),
```
Let me open a Github issue to track that the non-test code needs to throw an error for the below scenario
```
inpL3GwCfg: &L3GatewayConfig{
				Mode:   config.GatewayModeLocal,
				VLANID: &vlanid,
			},
```

Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:42]: FYI Opened issue https://github.com/ovn-org/ovn-kubernetes/issues/1478

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 4 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:177]: Screen Shot 2020-07-02 at 4 41 34 PM

From the coverage report, I am seeing that I am missing error path testing. I am guessing that's probably what you maybe referring to. I shall create the mocks for the Annotator interface in pkg/kube to test the error path

Comment (danwinship) [reply go-controller/pkg/util/node_annotations_unit_test.go:207]: well, I was thinking of `SetL3GatewayConfig` as being the function that did the marshalling, but I guess you effectively test that via the `MarshalJSON` test

Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:207]: > well, I was thinking of `SetL3GatewayConfig` as being the function that did the marshalling, but I guess you effectively test that via the `MarshalJSON` test

Yes, MarshalJSON() code path is 100% covered as part of the unit test
Screen Shot 2020-07-07 at 10 09 17 AM

Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:207]: Additionally, I have added tests to cover the error paths for the `SetL3GatewayConfig` function
Screen Shot 2020-07-07 at 10 14 32 AM

**Comment by vjayaramrh** [comment]
@danwinship I believe that I have addressed all your comments, please let me if there is any other issue that I need to take care that is blocking merge of this PR.

File: go-controller/pkg/util/node_annotations_unit_test.go [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+			desc:       "error: test unsupported gateway mode",
+			inputParam: []byte(`{"mode":"blah"}`),
+			errMatch:   fmt.Errorf("bad 'mode' value"),
+		},
+		{
+			desc:       "error: test bad VLANID input",
+			inputParam: []byte(`{"mode":"local","vlan-id":"A"}`),
+			errMatch:   fmt.Errorf("bad 'vlan-id' value"),
+		},
+
```
Comment (trozet) [inline go-controller/pkg/util/node_annotations_unit_test.go:115]: spurious newline

File: go-controller/pkg/util/node_annotations_unit_test.go [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+		{
+			desc:             "test error path when setting gateway annotation",
+			inpNodeAnnotator: mockAnnotator,
+			inputL3GwCfg:     L3GatewayConfig{},
+			onRetArgsAnnotatorList: []onCallReturnArgs{
+				{"Set", []string{"string", "interface{}"}, []interface{}{fmt.Errorf("mock error")}},
+			},
+		},
+		{
+			desc:             "success: apply empty Chassis id ",
```
Comment (trozet) [inline go-controller/pkg/util/node_annotations_unit_test.go:186]: extra trailing white space after Chassis id

File: go-controller/pkg/util/pod_annotation_unit_test.go [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+func TestMarshalPodAnnotation(t *testing.T) {
+	tests := []struct {
+		desc           string
+		inpPodAnnot    PodAnnotation
+		errAssert      bool  // used when an error string CANNOT be matched or sub-matched
+		errMatch       error //used when an error string CAN be matched or sub-matched
+		expectedOutput map[string]string
+	}{
+		{
+			desc:           "PodAnnotation instance with no fields set",
```
Comment (trozet) [inline go-controller/pkg/util/pod_annotation_unit_test.go:25]: you aligned the fields and values here, but in the subsequent entries they are not aligned

File: go-controller/pkg/util/pod_annotation_unit_test.go [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+				Routes: []PodRoute{
+					{
+						Dest:    ovntest.MustParseIPNet("0.0.0.0/0"),
+						NextHop: net.ParseIP("192.168.1.1"),
+					},
+				},
+			},
+		},
+		{
+			desc: "next hop not set for route ",
```
Comment (trozet) [inline go-controller/pkg/util/pod_annotation_unit_test.go:70]: extra space after route?

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 7 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:115]: shall fix

File: go-controller/pkg/util/node_annotations_unit_test.go (hunk 8 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/node_annotations_unit_test.go:186]: shall fix

File: go-controller/pkg/util/pod_annotation_unit_test.go (hunk 9 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/pod_annotation_unit_test.go:25]: gofmt when using the IDE automatically takes care of aligning them appropriately.
When I try to manually re-align them as can be seen below, I get an error when I execute `make gofmt`. This is because on Line31, a struct is being populated on mutli line

Screen Shot 2020-07-13 at 5 10 59 PM

File: go-controller/pkg/util/pod_annotation_unit_test.go (hunk 10 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/pod_annotation_unit_test.go:70]: shall fix
